File: api/src/main/java/org/apache/gravitino/semantic/Field.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.semantic;

import com.google.common.base.Preconditions;
import java.util.Arrays;
import java.util.Objects;
import javax.annotation.Nullable;
import org.apache.gravitino.annotation.Evolving;

/**
 * An immutable semantic field in a {@link Dataset}. A field gives a name to an {@link Expression}
 * and may carry dimension, display, type, AI, and extension metadata.
 */
@Evolving
public final class Field {

  private final String name;
  private final Expression expression;

  @Nullable private final Dimension dimension;
  @Nullable private final String label;
  @Nullable private final String description;
  @Nullable private final DataType datatype;
  @Nullable private final AIContext aiContext;
  @Nullable private final CustomExtension[] customExtensions;

  private Field(Builder builder) {
    this.name = builder.name;
    this.expression = builder.expression;
    this.dimension = builder.dimension;
    this.label = builder.label;
    this.description = builder.description;
    this.datatype = builder.datatype;
    this.aiContext = builder.aiContext;
    this.customExtensions =
        builder.customExtensions == null
            ? null
            : Arrays.copyOf(builder.customExtensions, builder.customExtensions.length);
  }

  /**
   * Creates a builder for an immutable {@link Field}.
   *
   * @return A new builder.
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Returns the field name.
   *
   * @return The field name.
   */
  public String name() {
    return name;
  }

  /**
   * Returns the expression that defines the field.
   *
   * @return The field expression.
   */
  public Expression expression() {
    return expression;
  }

  /**
   * Returns the dimension metadata for the field.
   *
   * @return The dimension metadata, or {@code null} if it is not set.
   */
  @Nullable
  public Dimension dimension() {
    return dimension;
  }

  /**
   * Returns the display label for the field.
   *
   * @return The field label, or {@code null} if it is not set.
   */
  @Nullable
  public String label() {
    return label;
  }

  /**
   * Returns the field description.
   *
   * @return The field description, or {@code null} if it is not set.
   */
  @Nullable
  public String description() {
    return description;
  }

  /**
   * Returns the semantic data type of the field.
   *
   * @return The semantic data type, or {@code null} if it is not set.
   */
  @Nullable
  public DataType datatype() {
    return datatype;
  }

  /**
   * Returns the AI context associated with the field.
   *
   * @return The AI context, or {@code null} if it is not set.
   */
  @Nullable
  public AIContext aiContext() {
    return aiContext;
  }

  /**
   * Returns a copy of the custom extensions associated with the field.
   *
   * @return The custom extensions, or {@code null} if they are not set.
   */
  @Nullable
  public CustomExtension[] customExtensions() {
    return customExtensions == null
        ? null
        : Arrays.copyOf(customExtensions, customExtensions.length);
  }

  /**
   * Compares this field with another object for value equality.
   *
   * @param other The object to compare with.
   * @return {@code true} if the objects are equal, otherwise {@code false}.
   */
  @Override
  public boolean equals(@Nullable Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof Field)) {
      return false;
    }
    Field that = (Field) other;
    return name.equals(that.name)
        && expression.equals(that.expression)
        && Objects.equals(dimension, that.dimension)
        && Objects.equals(label, that.label)
        && Objects.equals(description, that.description)
        && datatype == that.datatype
        && Objects.equals(aiContext, that.aiContext)
        && Arrays.equals(customExtensions, that.customExtensions);
  }

  /**
   * Returns the value-based hash code for this field.
   *
   * @return The hash code.
   */
  @Override
  public int hashCode() {
    return Objects.hash(
        name,
        expression,
        dimension,
        label,
        description,
        datatype,
        aiContext,
        Arrays.hashCode(customExtensions));
  }

  /**
   * Returns a string representation of this field.
   *
   * @return The string representation.
   */
  @Override
  public String toString() {
    return "Field{"
        + "name='"
        + name
        + '\''
        + ", expression="
        + expression
        + ", dimension="
        + dimension
        + ", label='"
        + label
        + '\''
        + ", description='"
        + description
        + '\''
        + ", datatype="
        + datatype
        + ", aiContext="
        + aiContext
        + ", customExtensions="
        + Arrays.toString(customExtensions)
        + '}';
  }

  /** A builder for immutable {@link Field} values. */
  public static final class Builder {

    private String name;
    private Expression expression;

    @Nullable private Dimension dimension;
    @Nullable private String label;
    @Nullable private String description;
    @Nullable private DataType datatype;
    @Nullable private AIContext aiContext;
    @Nullable private CustomExtension[] customExtensions;

    private Builder() {}

    /**
     * Sets the field name.
     *
     * @param name The non-empty field name.
     * @return This builder.
     */
    public Builder withName(String name) {
      this.name = name;
      return this;
    }

    /**
     * Sets the expression that defines the field.
     *
     * @param expression The field expression.
     * @return This builder.
     */
    public Builder withExpression(Expression expression) {
      this.expression = expression;
      return this;
    }

    /**
     * Sets the optional dimension metadata.
     *
     * @param dimension The dimension metadata, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withDimension(@Nullable Dimension dimension) {
      this.dimension = dimension;
      return this;
    }

    /**
     * Sets the optional display label.
     *
     * @param label The display label, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withLabel(@Nullable String label) {
      this.label = label;
      return this;
    }

    /**
     * Sets the optional field description.
     *
     * @param description The field description, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withDescription(@Nullable String description) {
      this.description = description;
      return this;
    }

    /**
     * Sets the optional semantic data type.
     *
     * @param datatype The semantic data type, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withDatatype(@Nullable DataType datatype) {
      this.datatype = datatype;
      return this;
    }

    /**
     * Sets the optional AI context.
     *
     * @param aiContext The AI context, or {@code null} to leave it unset.
     * @return This builder.
     */
    public Builder withAIContext(@Nullable AIContext aiContext) {
      this.aiContext = aiContext;
      return this;
    }

    /**
     * Sets the optional custom extensions.
     *
     * @param customExtensions The custom extensions, or {@code null} to leave them unset.
     * @return This builder.
     */
    public Builder withCustomExtensions(@Nullable CustomExtension[] customExtensions) {
      this.customExtensions = customExtensions;
      return this;
    }

    /**
     * Builds an immutable {@link Field}.
     *
     * @return The new field.
     * @throws IllegalArgumentException If the name is null or empty, the expression is null, or the
     *     custom extension array contains null.
     */
    public Field build() {
      Preconditions.checkArgument(
          name != null && !name.isEmpty(), "name must not be null or empty");
      Preconditions.checkArgument(expression != null, "expression must not be null");
      Preconditions.checkArgument(
          customExtensions == null || Arrays.stream(customExtensions).allMatch(Objects::nonNull),
          "customExtensions must not contain null");
      return new Field(this);
    }
  }
}

File: clients/client-python/gravitino/api/semantic/field.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from typing import Optional

from gravitino.api.semantic.ai_context import AIContext
from gravitino.api.semantic.custom_extension import CustomExtension
from gravitino.api.semantic.data_type import DataType
from gravitino.api.semantic.dimension import Dimension
from gravitino.api.semantic.expression import Expression
from gravitino.api.semantic.semantic_utils import check_no_none_elements
from gravitino.utils.precondition import Precondition


class Field:  # pylint: disable=too-many-instance-attributes
    """A named expression exposed by a Semantic Model dataset.

    Field names are unique within their dataset.
    """

    def __init__(
        self,
        name: str,
        expression: Expression,
        dimension: Optional[Dimension] = None,
        label: Optional[str] = None,
        description: Optional[str] = None,
        datatype: Optional[DataType] = None,
        ai_context: Optional[AIContext] = None,
        custom_extensions: Optional[list[CustomExtension]] = None,
    ):
        Precondition.check_argument(
            name is not None and name != "", "name must not be null or empty"
        )
        Precondition.check_argument(
            expression is not None, "expression must not be null"
        )
        check_no_none_elements("customExtensions", custom_extensions)

        self._name = name
        self._expression = expression
        self._dimension = dimension
        self._label = label
        self._description = description
        self._datatype = datatype
        self._ai_context = ai_context
        self._custom_extensions = (
            None if custom_extensions is None else list(custom_extensions)
        )

    def name(self) -> str:
        """Returns the field name."""
        return self._name

    def expression(self) -> Expression:
        """Returns the expression that produces the field."""
        return self._expression

    def dimension(self) -> Optional[Dimension]:
        """Returns the dimension marker, or `None` if the field is not a dimension."""
        return self._dimension

    def label(self) -> Optional[str]:
        """Returns the display label, or `None` if it is not set."""
        return self._label

    def description(self) -> Optional[str]:
        """Returns the field description, or `None` if it is not set."""
        return self._description

    def datatype(self) -> Optional[DataType]:
        """Returns the logical data type, or `None` if it is not set."""
        return self._datatype

    def ai_context(self) -> Optional[AIContext]:
        """Returns the AI context, or `None` if it is not set."""
        return self._ai_context

    def custom_extensions(self) -> Optional[list[CustomExtension]]:
        """Returns the custom extensions, or `None` if they are not set."""
        return (
            None if self._custom_extensions is None else list(self._custom_extensions)
        )

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Field):
            return False
        return (
            self._name == other.name()
            and self._expression == other.expression()
            and self._dimension == other.dimension()
            and self._label == other.label()
            and self._description == other.description()
            and self._datatype == other.datatype()
            and self._ai_context == other.ai_context()
            and self._custom_extensions == other.custom_extensions()
        )

    def __hash__(self) -> int:
        return hash(
            (
                self._name,
                self._expression,
                self._dimension,
                self._label,
                self._description,
                self._datatype,
                self._ai_context,
                (
                    None
                    if self._custom_extensions is None
                    else tuple(self._custom_extensions)
                ),
            )
        )

    def __repr__(self) -> str:
        return (
            f"Field(name={self._name!r}, expression={self._expression!r}, "
            f"dimension={self._dimension!r}, label={self._label!r}, "
            f"description={self._description!r}, datatype={self._datatype!r}, "
            f"aiContext={self._ai_context!r}, "
            f"customExtensions={self._custom_extensions!r})"
        )
